src/Analyzers/CSharp/Analyzers/UseUTF8StringLiteral/UseUTF8StringLiteralDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,204 @@
+        {
+            // When the first elements parent is as argument, or an edge case for collection
+            // initializers where the Add method takes a param array, it means we have a parameter array.
+            // We raise the diagnostic on all of the parameters that make up the array. We could do just
+            // the first element, but that might be odd seeing: M(1, 2, [|3|], 4, 5)
+            var span = TextSpan.FromBounds(elements[0].Syntax.SpanStart, elements[^1].Syntax.Span.End);
+            var location = Location.Create(syntaxNode.SyntaxTree, span);
+
+            ReportDiagnostic(context, syntaxNode, severity, location, operationLocation);
+        }
+
+        private void ReportArrayCreationDiagnostic(OperationAnalysisContext context, SyntaxNode syntaxNode, ReportDiagnostic severity)
+        {
+            // When the user writes the array creation we raise the diagnostic on the first token, which will be the "new" keyword
+            var location = syntaxNode.GetFirstToken().GetLocation();
+
+            ReportDiagnostic(context, syntaxNode, severity, location, ArrayCreationOperationLocation.Current);
+        }
+
+        private void ReportDiagnostic(OperationAnalysisContext context, SyntaxNode syntaxNode, ReportDiagnostic severity, Location location, ArrayCreationOperationLocation operationLocation)
+        {
+            // Store the original syntax location so the code fix can find the operation again
+            var additionalLocations = ImmutableArray.Create(syntaxNode.GetLocation());
+
+            // Also let the code fix where to look to find the operation that originally trigger this diagnostic
+            var properties = ImmutableDictionary<string, string?>.Empty.Add(nameof(ArrayCreationOperationLocation), operationLocation.ToString());
+
+            context.ReportDiagnostic(
+                DiagnosticHelper.Create(Descriptor, location, severity, additionalLocations, properties));
+        }
+
+        internal static bool TryConvertToUTF8String(StringBuilder? builder, ImmutableArray<IOperation> arrayCreationElements)
+        {
+            for (var i = 0; i < arrayCreationElements.Length;)
+            {
+                // Need to call a method to do the actual rune decoding as it uses stackalloc, and stackalloc
+                // in a loop is a bad idea
+                if (!TryGetNextRune(arrayCreationElements, i, out var rune, out var bytesConsumed))
+                    return false;
+
+                i += bytesConsumed;
+
+                if (builder is not null)
+                {
+                    if (rune.TryGetEscapeCharacter(out var escapeChar))
+                    {
+                        builder.Append('\\');
+                        builder.Append(escapeChar);
+                    }
+                    else
+                    {
+                        builder.Append(rune.ToString());
```

Comment:
there's no nicer way of appending a rune to an arraybuilder? :'(   seems unfortunate to have to allocate for this purpose.

Reply:
Nothing nice, no. I _think_ I could stack alloc a char array, and write to that as a span, but then I have to call ToArray on the span to append it to the string builder anyway. And I'd have to do it in a separate method because this is in a loop